### PR TITLE
Storage

### PR DESCRIPTION
**Pull request by dilyn-corner**
This is my updated version of @mcpcpc's storage page. 

The work mcpcpc provided me was a fantastic starting point, and I kept true to the structure for most of the page. Clarifications on pain-points regarding what exactly partitions amount to and the purpose they serve were a primary focus I had, along with efficiency and common-use cases. 
My biggest goal is to refrain from merely reproducing the Arch Wiki page. This information should be targeted, clear, and eliminate confusion (my biggest complaint against The Bible(tm)). I believe this page accomplishes that goal as it stands. 

There are currently two major issues I have with this PR:
1) two pages that do not currently exist in the wiki are referenced (zswap, zram)
2) there is no `fstab` section

I plan on writing articles for zswap and zram soon (this weekend/this week), and they may or may not be consolidated into a single article (they're similar, but distinct. Makes it tricky). 
I would like to have a wiki entry covering the `fstab` as it is a frequently misunderstood file by even long-time Linux users. At the very least, an example `fstab` would be useful (considering we lack a `genfstab` tool). 

The reason for this PR given the two aforementioned halting points is simply to generate feedback and comb through the grammar while I resolve these issues :)

### CODE REVIEW

File: boot/storage.txt [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+    |                  |   filesystem is the structure for a given partition.  |
+    |                  |                                                       |
+    +------------------+-------------------------------------------------------+
+
+    Some useful facts and things to keep in mind: 
+
+    Due to using 32-bit identifiers, DOS partitioning tables cannot handle disks
+    that are >2 TBs in size. 
+    
+    Unless a extended partition is created, DOS disklables supports a maximum of
```
Comment (dylanaraps) [inline boot/storage.txt:109]: `a extended` -> `an extended`

File: boot/storage.txt [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+    Firmware Interface or the Basic Input/Output System. These is determined
+    strictly by the motherboard used in a given system. The latter, however,
+    refer to particular partition types: GUID Partition Tables versus the
+    Master Boot Record. These types directly limit the structures available
+    on any given disk drive - both the number of partitions available, as 
+    well as the size of any given partition.
+    
+    An important point of consideration is whether or not a separate HOME 
+    partition should be used. This is useful in cases where users wish to keep
+    there data in a location separate from their operating system, for instance
```
Comment (dylanaraps) [inline boot/storage.txt:146]: `there data` -> `their data`.

File: boot/storage.txt [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+    - cfdisk: a partition editor with a curses-based UI, part of util-linux [4]
+    - parted: a GNU partition editor [5]
+
+    It is also possible to use the live CD version of parted, GParted [6], to 
+    partition disks, and it offers many other useful features for disk 
+    management and data recovery.
+
+    In general, most programs can handle the majority of use-cases. The tool of 
+    choice is almost entirely user preference. However, some utilities have 
+    powerful and advanced features that can render data on disks damaged. fdisk
```
Comment (dylanaraps) [inline boot/storage.txt:181]:
```
some utilities have powerful and advanced features that can render data on disks damaged
```

```
some utilities have powerful and advanced features that (if used incorrectly or without caution) can cause irrecoverable damage to disk data.
```

This may be better though I'm still not happy with it. :P

File: boot/storage.txt [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+    is a one such program, and is useful in cases where more fine-grained
+    control is required. For general system setup, however, any tool will do.
+    
+    
+[2.0] Partitioning Schemata
+________________________________________________________________________________
+
+Because disk setup is in general difficult to modify after a particular
```
Comment (dylanaraps) [inline boot/storage.txt:189]:
```
Great care should be taken when picking a partition structure as changing it afterwards is (generally) a difficult endeavor. Ensure that the chosen structure meets your use case. ...
```

Just a suggestion/idea (I don't like sentences starting with "because"). Feel free to reword this as you wish.

File: boot/storage.txt [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+    as close to the first partition as possible, that is at least half the
+    amount of RAM available - this number is a sliding scale and varies by use
+    case. For instance, if few RAM intensive programs are going to be used, and 
+    little compiling will be done on the host system, less swap space is needed
+    (if any at all). 
+
+    Historically, swap was placed at the beginning of disks as this allowed for 
+    the fastest possible seek times on traditional spinning disk drives (hard
+    drives). This was preferred because disk access times are orders of 
+    magnitude slower than RAM access time, and long swap operations  result in
```
Comment (dylanaraps) [inline boot/storage.txt:255]: `operations  result` -> `operations result` (two spaces -> one space)

File: boot/storage.txt [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+    happening. With access to these newer technologies, users can simply create
+    a file and define it to be a dedicated swap location which is used 
+    identically to the traditional swap partition. This has left the need for a
+    swap partition largely to cases where solid state storage is not an option.
+    
+    For more modern systems, swap files are in general preferable to partitions.
+    Care should be taken, however. Changes may occur outside of the user's
+    control that necessitate intervention to allow the system to boot. For
+    instance, if a swapfile disappears but is still referenced in /etc/fstab
+    will drop a system into an emergency shell upon boot. Additionally,
```
Comment (dylanaraps) [inline boot/storage.txt:274]:
```
For instance, if a swapfile disappears but is still referenced in /etc/fstab will drop a system into an emergency shell upon boot.
```

```
For instance, if a swapfile disappears but is still referenced in /etc/fstab, disk mount at boot will fail and the system will drop into an emergency shell.
```

Can

File: boot/storage.txt [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+    DOS, NTFS) filesystems. All of these have built-in kernel support.
+
+    For a more complete list of other filesystem options and what limitations 
+    and features exist for them, refer to the Arch Wiki [9].
+
+
+    [3.2] Swap Partition
+    ____________________________________________________________________________ 
+    
+    To set up a partition as a Linux swap space, the mkswap command is use.
```
Comment (dylanaraps) [inline boot/storage.txt:332]: `is use` -> `is used`

File: boot/storage.txt [hunk 8]
```diff
@@ -0,0 +1,485 @@
...
+    |                                                                          |
+    |   mkswap /dev/sdXY                                                       |
+    |   swapon /dev/sdXY                                                       |
+    |                                                                          |
+    +--------------------------------------------------------------------------+  
+
+
+    [3.3] Swap file
+    ____________________________________________________________________________    
+    As an alternative to creating an entire partition, a swap file offers
```
Comment (dylanaraps) [inline boot/storage.txt:346]: Add blank line above this one.

File: boot/storage.txt [hunk 9]
```diff
@@ -0,0 +1,485 @@
...
+    +--------------------------------------------------------------------------+
+
+    $LOCATION is the partition where the swapfile should be located. For the 
+    next few commands, assume $LOCATION=/swapfile.
+
+    $BYTES is The block size, which is determined by the filesystem on the 
+    partition where $LOCATION is. This setting is important for optimizing
+    read/write throughput. dd defaults to 512 bytes, but this is suboptimal for 
+    drives with larger supported block sizes. To determine the supported block
+    size for your device, run dumpe2fs (provided by e2fsprogs):
```
Comment (dylanaraps) [inline boot/storage.txt:366]: iirc there's some `/sys` or `/proc` file containing this information. I could be wrong though. I'll look into it.

File: boot/storage.txt [hunk 10]
```diff
@@ -0,0 +1,485 @@
...
+    +--------------------------------------------------------------------------+
+    |                                                                          |
+    |   $ dumpe2fs -h /dev/sdXY | grep 'Block size'                            |
+    |                                                                          |
+    +--------------------------------------------------------------------------+
+
+    For instance, if the result is 4096, bs=4k should be chosen. 
+    
+    $SIZE is the space you wish to allocate to the swapfile, suffixed by
+    kilobytes (kB), kibibytes (k), megabytes (M), mebibytes (MB), etc.
```
Comment (dylanaraps) [inline boot/storage.txt:377]: Make all units uppercase? Are they case-sensitive in the first place? If not, it's probably nice to use consistent casing here.

File: boot/storage.txt [hunk 11]
```diff
@@ -0,0 +1,485 @@
...
+
+    +--------------------------------------------------------------------------+
+    |                                                                          |
+    |   $ chmod 600 /swapfile                                                  |
+    |   $ mkswap    /swapfile                                                  |
+    |   $ swapon    /swapfile                                                  |
+    |                                                                          |
+    +--------------------------------------------------------------------------+
+
+    To remove a swap file, it's a simple as disabling and deleting the it,
```
Comment (dylanaraps) [inline boot/storage.txt:389]: `it's a` -> `it's as`

Comment (dylanaraps) [reply boot/storage.txt:389]: `the it` -> `it`

File: boot/storage.txt [hunk 12]
```diff
@@ -0,0 +1,485 @@
...
+    |                                                                          |
+    +--------------------------------------------------------------------------+
+    
+
+[4.0] Partitioning Example
+________________________________________________________________________________
+
+The following is a step-by-step partitioning example. 
+
+PLEASE BE AWARE: the following commands can cause irreparable damage to the
```
Comment (dylanaraps) [inline boot/storage.txt:404]: `the` -> `The`

**Comment by dylanaraps** [comment]
Looking good. Left some nitpicks for now. Will (probably) send more substantial comments when I can spend more time looking at this.

File: boot/storage.txt (hunk 3 above)
Comment (dilyn-corner) [reply boot/storage.txt:181]: It's not even what I intended to say xD
1) Don't draft multiple sections at once
2) Don't drink and draft

File: boot/storage.txt (hunk 4 above)
Comment (dilyn-corner) [reply boot/storage.txt:189]: 'Because' is a dirty word and I should know better.

File: boot/storage.txt (hunk 8 above)
Comment (dilyn-corner) [reply boot/storage.txt:346]: Do you mean there's an extra blank at 345?

File: boot/storage.txt (hunk 9 above)
Comment (dilyn-corner) [reply boot/storage.txt:366]: There should be; there's a way to make e.g. `busybox` print this information too. I'll find it. 

I had never actually heard of dumpe2fs so I was very excited to include it.

File: boot/storage.txt (hunk 10 above)
Comment (dilyn-corner) [reply boot/storage.txt:377]: They are case sensitize. `dd` is a dumb and broken command.
What other program differentiates between megabytes and mebibytes ffs

**Comment by dilyn-corner** [comment]
Nitpicks are important if only because I miss small things when I write too much. Some solid suggestions for things to consider were sprinkled in as a bonus!

File: boot/storage.txt (hunk 8 above)
Comment (dylanaraps) [reply boot/storage.txt:350]:
```
    ____________________________________________________________________________    
    As an alternative to creating an entire partition, a swap file offers 
```

```
    ____________________________________________________________________________    

    As an alternative to creating an entire partition, a swap file offers 
```

Comment (dilyn-corner) [reply boot/storage.txt:351]: I see it now! It seems that the dividing line was using hyphens instead of underscores? I've resolved this.

How are your eyes so keen.

**Comment by dilyn-corner** [comment]
Still unable to find a easy way to read blocksize from /proc, but `stat` is an option that is probably better than `e2fsprogs` - how uniform are `stat` implementations, anyways...

An alternative is to just `df`, `touch test`, `df` and find the difference between the given blocks available :laughing:

File: boot/storage.txt [hunk 13]
```diff
@@ -0,0 +1,567 @@
...
+    choice is almost entirely user preference. However, some utilities have 
+    powerful and advanced features that can aid in fixing damaged disks. fdisk
+    is a one such program, and is useful in cases where more fine-grained
+    control is required. For general system setup, however, any tool will do.
+    
+    
+[1.0] Partitioning Schemata
+________________________________________________________________________________
+
+Great care should be taken when pkcing a partition structure, as changing it
```
Comment (nerditup) [inline boot/storage.txt:191]: pkcing --> picking 👍

**Comment by dilyn-corner** [comment]
I don't think boot/ is the proper place for this page. 

I would like to hold this PR after I 1) finish the zram and zswap articles, and 2) determine where this article should go. I think technically it belongs in a new section (with the two aforementioned articles). I think this article is not of the same type as the other articles in boot/ - for instance, I would expect a grub article to appear here, not a disk-setup article.

I will be working on a zswap/zram article this week (famous last words).

**Comment by dilyn-corner** [comment]
I've made some preliminary pages for zswap and zram. I don't see any reason to deny them from this PR besides keeping individual pages to their own requests; I can separate them out if you would like @dylanaraps. 

I'm going to have to test zram a bit more before I feel comfortable merging the page but aside from that I think the information presented gets at the essence of the features; they're quite minimalist and there isn't a whole lot of configuration outside of general setup. 

Feel free to leave thoughts, comments, and concerns :)

File: storage/zswap.txt [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+to your kernel command line (via the bootloader or the built-in command line) to
+have zswap enabled at boot-time. To enable zswap at runtime,
+
++------------------------------------------------------------------------------+
+|                                                                              |
+|   $ echo 1 > /sys/module/zswap/parameters/enabled                            |
+|                                                                              |
++------------------------------------------------------------------------------+
+
+Zswap has many options that can also be configured at runtime, including which
```
Comment (dylanaraps) [inline storage/zswap.txt:47]: The zram pages don't capitalize `zram` but the zswap page does capitalize `zswap`. `Zswap` -> `zswap` to match the zram page?

File: storage/zswap.txt [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+| zsmalloc    | Designed for zram - better under low memory conditions         |
+|             |                                                                |
++-------------+----------------------------------------------------------------+
+
+In general, z3fold should be preferred to zbud; the latter is supported solely
+for compatibility purposes. z3fold provides a better compression ratio and
+should be preferred when possible. 
+
+zsmalloc has a very different page allocation method than either zbud or z3fold,
+and provides for greater storage density. However, zsmalloc does not implement
```
Comment (dylanaraps) [inline storage/zswap.txt:79]: `for` -> `far` ?

Comment (dylanaraps) [reply storage/zswap.txt:79]: This may be correct already though... both make sense. :P

File: storage/zram.txt [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+|                                                                              |
+|   # create zram devices of $SIZE 2G                                          |
+|   # Mark the devices as swap devices                                         |
+|   # Enable the swap devices                                                  |
+|                                                                              |
+|   for dev in 0 1 2 3; do                                                     |
+|       echo 2G > /sys/block/zram$dev/disksize                                 |
+|       mkswap /dev/zram$dev                                                   |
+|       swapon /dev/zram$dev -p 10                                             |
+|   done                                                                       |
```
Comment (dylanaraps) [inline storage/zram.txt:65]: I know this is a little pedantic but these variables should be quoted. `"/dev/zram$dev"`. Same for the second block of code.

**Comment by dylanaraps** [comment]
Once these nits are fixed I'm very happy to merge. Thanks once again for doing this. :)

File: storage/zswap.txt (hunk 15 above)
Comment (dilyn-corner) [reply storage/zswap.txt:79]: They do both make sense; if 'provides for' is too pretentious sounding I can change it XD

File: storage/zram.txt (hunk 16 above)
Comment (dilyn-corner) [reply storage/zram.txt:65]: Pedantry is my bread and butter.

**Comment by dilyn-corner** [comment]
It was my pleasure! I learned a fair bit (and will now be utilizing `zram` as a result of this work!)

I'm a huge fan of proper and correct documentation - any little bit I can contribute to this project makes my heart a little warmer.
